File: pt-1/3855.sum-of-k-digit-numbers-in-a-range.go
package leetcode

//
// @lc app=leetcode id=3855 lang=golang
//
// [3855] Sum of K-Digit Numbers in a Range
//
// Notes
// The sum of chosen digits from [l,r] is the arithmetic-series sum. For every
// position, the other k-1 positions can be chosen independently in count^(k-1)
// ways; multiplying by the repunit 111..1 places the same digit contribution in
// every decimal position. Modular inverse of 9 builds the repunit. Time:
// O(log k). Space: O(1).
//
// @lc code=start

const mod3855 int64 = 1_000_000_007

func SumOfNumbers3855(l int, r int, k int) int {
	count := int64(r - l + 1)
	digitSum := int64(l+r) * count / 2
	otherPositions := powMod3855(count%mod3855, int64(k-1))
	repunit := (powMod3855(10, int64(k)) - 1 + mod3855) % mod3855
	repunit = repunit * powMod3855(9, mod3855-2) % mod3855
	return int(digitSum % mod3855 * otherPositions % mod3855 * repunit % mod3855)
}

func powMod3855(base int64, exp int64) int64 {
	result := int64(1)
	base %= mod3855
	for exp > 0 {
		if exp&1 == 1 {
			result = result * base % mod3855
		}
		base = base * base % mod3855
		exp >>= 1
	}
	return result
}

// @lc code=end

File: pt-1/3855.sum-of-k-digit-numbers-in-a-range.java
/*
 * @lc app=leetcode id=3855 lang=java
 *
 * [3855] Sum of K-Digit Numbers in a Range
 *
 * Every position independently chooses one number from [l, r]. The sum of all
 * chosen digits over all k-length numbers is arithmeticSum(l..r) *
 * count^(k-1) repeated in every decimal position, hence multiply by the
 * k-digit repunit.
 *
 * Time: O(log k). Space: O(1).
 */

// @lc code=start
class Solution {
    private static final long MOD = 1_000_000_007L;

    public int sumOfNumbers(long l, long r, int k) {
        long count = r - l + 1;
        long digitSum = ((l + r) % MOD) * (count % MOD) % MOD * inv2() % MOD;
        long otherPositions = powMod(count % MOD, k - 1);
        long repunit = (powMod(10, k) - 1 + MOD) % MOD * powMod(9, MOD - 2) % MOD;
        return (int) (digitSum * otherPositions % MOD * repunit % MOD);
    }

    private long inv2() {
        return (MOD + 1) / 2;
    }

    private long powMod(long base, long exp) {
        long result = 1;
        base %= MOD;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = result * base % MOD;
            }
            base = base * base % MOD;
            exp >>= 1;
        }
        return result;
    }
}
// @lc code=end
